pkg/store/cache/caching_bucket.go

```diff
@@ -132,35 +131,184 @@ func (cb *CachingBucket) ReaderWithExpectedErrs(expectedFunc objstore.IsOpFailur
 	return cb.WithExpectedErrs(expectedFunc)
 }
 
-var chunksMatcher = regexp.MustCompile(`^.*/chunks/\d+$`)
+func (cb *CachingBucket) Iter(ctx context.Context, dir string, f func(string) error) error {
+	cfgName, cfg := cb.cfg.findIterConfig(dir)
+	if cfg == nil {
+		return cb.Bucket.Iter(ctx, dir, f)
+	}
+
+	cb.operationRequests.WithLabelValues(opIter, cfgName).Inc()
+
+	key := cachingKeyIter(dir)
+	data := cfg.cache.Fetch(ctx, []string{key})
+	if data[key] != nil {
+		list, err := cfg.codec.Decode(data[key])
+		if err == nil {
+			cb.operationHits.WithLabelValues(opIter, cfgName).Inc()
+			for _, n := range list {
+				if err := f(n); err != nil {
+					return err
+				}
+			}
+			return nil
+		}
+		level.Warn(cb.logger).Log("msg", "failed to decode cached Iter result", "err", err)
+	}
+
+	// Iteration can take a while (esp. since it calls function), and iterTTL is generally low.
+	// We will compute TTL based time when iteration started.
+	iterTime := time.Now()
+	var list []string
+	err := cb.Bucket.Iter(ctx, dir, func(s string) error {
+		list = append(list, s)
+		return f(s)
+	})
+
+	remainingTTL := cfg.ttl - time.Since(iterTime)
+	if err == nil && remainingTTL > 0 {
+		if data, err := cfg.codec.Encode(list); err == nil {
+			cfg.cache.Store(ctx, map[string][]byte{key: data}, remainingTTL)
+		} else {
```

Comment:
As you seen I love explicitlness and don't like `else` that's why I would suggest returning `nil` explicitly instead of else's (:

Reply:
Of course feel free to disagree

Reply:
```		if data, err := cfg.codec.Encode(list); err == nil {
			cfg.cache.Store(ctx, map[string][]byte{key: data}, remainingTTL)
			return nil
		}
		level.Warn(cb.logger).Log("msg", "failed to encode Iter result", "err", err)
```

version would have a subtle bug -- using of incorrect `err` variable. But I can make it bit more explicit.

Reply:
:+1: Up to you, sorry for being a pain (: